fastavro/_read_py.py

```diff
@@ -523,50 +558,63 @@ def acquaint_schema(schema):
 
 def _iter_avro(fo, header, codec, writer_schema, reader_schema):
     """Return iterator over avro records."""
+    for block in _iter_avro_blocks(fo, header, codec, writer_schema,
+                                   reader_schema):
+        for i in range(block.num_records):
+            yield read_data(block.bytes, writer_schema, reader_schema)
+
+
+def _iter_avro_blocks(fo, header, codec, writer_schema, reader_schema):
+    """Return iterator over avro blocks."""
     sync_marker = header['sync']
-    # Value in schema is bytes
 
     read_block = BLOCK_READERS.get(codec)
     if not read_block:
         raise ValueError('Unrecognized codec: %r' % codec)
 
-    block_count = 0
     while True:
-        block_count = read_long(fo, None)
-        block_fo = read_block(fo)
+        offset = fo.position()
+        num_block_records = read_long(fo)
 
-        for i in xrange(block_count):
-            yield read_data(block_fo, writer_schema, reader_schema)
+        block_bytes = read_block(fo)
 
         skip_sync(fo, sync_marker)
 
+        size = fo.position() - offset
 
-class reader:
-    """Iterator over avro file.
+        yield Block(
+            block_bytes, num_block_records, codec, reader_schema,
+            writer_schema, offset, size
+        )
 
-    Parameters
-    ----------
-    fo: file-like
-        Input stream
-    reader_schema: dict, optional
-        Reader schema
 
+class Block:
+    def __init__(self, bytes, num_records, codec, reader_schema, writer_schema,
+                 offset, size):
+        self.bytes = bytes
+        self.num_records = num_records
+        self.codec = codec
+        self.reader_schema = reader_schema
+        self.writer_schema = writer_schema
+        self.offset = offset
+        self.size = size
 
+    def __iter__(self):
+        for i in range(self.num_records):
+            yield read_data(self.bytes, self.writer_schema,
+                            self.reader_schema)
 
-    Example::
+    def __str__(self):
+        return ("Avro block: %d bytes, %d records, codec: %s, position %d+%d"
+                % (len(self.bytes), self.num_records, self.codec, self.offset,
+                   self.size))
 
-        from fastavro import reader
-        with open('some-file.avro', 'rb') as fo:
-            avro_reader = reader(fo)
-            schema = avro_reader.schema
-            for record in avro_reader:
-                process_record(record)
-    """
 
+class file_reader:
     def __init__(self, fo, reader_schema=None):
-        self.fo = fo
         try:
-            self._header = read_data(fo, HEADER_SCHEMA)
+            self._fo_reader = FileObjectReader(fo)
```

Comment:
Can we move this to before the `try`? Just to make it clear that the `try` is only for the `read_data` call?